diff --git a/pkgs/applications/audio/librespot/default.nix b/pkgs/applications/audio/librespot/default.nix
index f5ff221e1bfe..b85f6b53af1d 100644
--- a/pkgs/applications/audio/librespot/default.nix
+++ b/pkgs/applications/audio/librespot/default.nix
@@ -25,13 +25,18 @@ rustPlatform.buildRustPackage rec {
   version = "0.6.0";
 
   src = fetchFromGitHub {
-    owner = "librespot-org";
+    owner = "googlebot42";
     repo = "librespot";
-    rev = "v${version}";
-    sha256 = "sha256-dGQDRb7fgIkXelZKa+PdodIs9DxbgEMlVGJjK/hU3Mo=";
+    rev = "ab9ecb6d936edf24b18258be981966692e705a02";
+    sha256 = "sha256-7Wx7+2OYAczNKzqRH2v8fwNAfll27Bdr3Fo23uDtQ4Y=";
   };
 
-  cargoHash = "sha256-SqvJSHkyd1IicT6c4pE96dBJNNodULhpyG14HRGVWCk=";
+  cargoHash = "sha256-s6qUI+pDyfBRhuPuBsi2eVxH9DLXEgISf/N4ZY3zDgc=";
+
+  cargoBuildFlags = [
+    "--features"
+    "passthrough-decoder"
+  ];
 
   nativeBuildInputs = [
     pkg-config
